crates/client/rpc-core/src/lib.rs

```diff
@@ -35,17 +38,56 @@ pub struct FunctionCall {
     pub calldata: Vec<FieldElement>,
 }
 
+// In order to mix tagged and untagged {de}serialization for BlockId (see starknet RPC standard)
+// in the same object, we need this kind of workaround with intermediate types
+
+#[derive(Serialize, Deserialize, Clone)]
+enum BlockIdTaggedVariants {
+    #[serde(rename = "block_hash")]
+    BlockHash(BlockHash),
+    #[serde(rename = "block_number")]
+    BlockNumber(BlockNumber),
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(untagged)]
+enum BlockIdUntagged {
+    Tagged(BlockIdTaggedVariants),
+    BlockTag(BlockTag),
+}
+
 /// A block hash, number or tag
-/// TODO: fix block_tag
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[serde(from = "BlockIdUntagged")]
+#[serde(into = "BlockIdUntagged")]
 pub enum BlockId {
-    #[serde(rename = "block_hash")]
     BlockHash(BlockHash),
-    #[serde(rename = "block_number")]
     BlockNumber(BlockNumber),
     BlockTag(BlockTag),
 }
 
+impl From<BlockIdUntagged> for BlockId {
+    fn from(value: BlockIdUntagged) -> Self {
+        match value {
+            BlockIdUntagged::Tagged(v) => match v {
+                BlockIdTaggedVariants::BlockHash(h) => Self::BlockHash(h),
+                BlockIdTaggedVariants::BlockNumber(n) => Self::BlockNumber(n),
+            },
+            BlockIdUntagged::BlockTag(t) => Self::BlockTag(t),
+        }
+    }
+}
+
+impl From<BlockId> for BlockIdUntagged {
+    fn from(value: BlockId) -> Self {
```

Comment:
add rust doc

Reply:
C'est le trait from.
Ya pas grand chose que je peux ajouter avec des mots qui ne soit déjà dit par la doc du trait from et par la lecture de l'implem

Reply:
I see your point but still I want to soon enforce Rust doc everywhere through CI.
Even if it's obvious I would still add some Rust doc. Even if there is doc in a trait, implementations can always have noteworthy information that are specific to them